[WIP] FeatherstonesMethod

jbm950 opened the pull request:
Created an example of how Featherstone's method would be setup/called in the docstring of `FeatherstonesMethod` (I renamed it properly locally).

Thoughts @moorepants? What are some areas to target for tomorrow's meeting? I find it kind of funny that simbody creates the same system in significantly fewer lines despite being c++. I'm using Sahil's joints and body classes PR #9809

sahilshekhawat committed:
Added all specific joints class

sahilshekhawat committed:
Fixed tests and some minor bugs in Joints classes.

sahilshekhawat committed:
Aded tests for planar joint and spherical joint

sahilshekhawat committed:
Merge branch 'master' of https://github.com/sympy/sympy into gsoc15-joints-class

sahilshekhawat committed:
Fixed minor issues to make all tests pass after merging Body class

sahilshekhawat committed:
Fixed doctests and added Joints to docs

sahilshekhawat committed:
Fixed test failures in joints

sahilshekhawat committed:
Added arguments tests for each joint

sahilshekhawat committed:
Added more clear docstrings and changed implementation of planar joint

sahilshekhawat committed:
Fixed doctests and changed planar joint's implementation

sahilshekhawat committed:
Fixed spherical JOint implementation

jbm950 committed:
Merge branch 'gsoc15-joints-class' of https://github.com/sahilshekhawat/sympy into joints

jbm950 committed:
First commit for FeatherstonesMethod

This commit creates a new file and class for the development of
featherstones method. Currently the docstrings contain just an example
of how featherstone's method would be setup and executed.

moorepants commented on sympy/physics/mechanics/featherstone.py:
```diff
@@ -0,0 +1,78 @@
+        ...                   parent_axis=bodyA.frame.z,
+        ...                   child_axis=bodyB.frame.z)
+
+        >>> featherstone = FeatherstonesMethod(ground)
```
It'd be nice to see examples of some of the methods this class will have, with expected output.

moorepants commented on sympy/physics/mechanics/featherstone.py:
```diff
@@ -0,0 +1,78 @@
+        >>> bodyA = Body('A', mass_center_A, mA, frameA, body_A_inertia_dyadic)
+        >>> bodyB = Body('B', mass_center_B, mB, frameB, body_B_inertia_dyadic)
+        >>> ground_body = Body('G', ground_point, 0, ground,
+        ...                    ground_inertia_dyadic)
```
I think Body was supposed to create all of the points, scalars, etc if you don't supply them, that would reduce some of the code to be more similar to simbody (simbody users just pass in numerical values for all these things).

jbm950 replied:
There's some lack of clarity in `Body`'s docstring as to when it would fill in information for a body vs. a particle. Also there's an if statement that appears twice in its `__init__()` method where code is being run that wouldn't be used when the second if statement is run. I'm going to put together a PR to alter the `Body` class and link it here.

jbm950 replied:
The PR can now be found in PR #11416

moorepants commented on sympy/physics/mechanics/featherstone.py:
```diff
@@ -0,0 +1,78 @@
+        ...                   parent_point_pos=(0, 0, 0),
+        ...                   child_point_pos=(0, 1, 0),
+        ...                   parent_axis=bodyA.frame.z,
+        ...                   child_axis=bodyB.frame.z)
```
What kinds of methods on joint will we expect to see for use in the order n method? Maybe sketch these out so we can get an idea of them.

moorepants commented on sympy/physics/mechanics/featherstone.py:
```diff
@@ -0,0 +1,78 @@
+
+        TODO: Should the class simply make Denavit-Hartenberg parameters for all
+        of the joints where num_joints=num_bodies for an open chain system?
+        coordinates : ordered iterables of functions of time
```
I don't think we need Denavit-Hartenberg parameters. Isn't that a limited set of joints? Or can their parameters describe any joint? Is this what Featherstone uses?

jbm950 committed:
pushing for discussion purposes

jbm950 commented:
@moorepants 

I've prototyped what Featherstone's algorithm on page 132 might look like in python (all three passes are commented out). I've also prototyped what it would look like to build the kinematic tree info from the base body. Lastly I've prototyped in two methods what it would look like if the kinematic tree is built before the algorithm and what it looks like if the kinematic tree is built during the first pass of the algorithm.

jbm950 committed:
Created a mock up of pass 2 and 3

A mock up of pass 2 and 3 was added to the method demonstrating the
separated pass 1.

jbm950 committed:
Merge branch 'joints' into featherstone

jbm950 committed:
Reduced the joints to 2

I removed the code, documentation and tests for all of the joints except
the revolute and prismatic joints because these are the two I plan to
use in Featherstone's method. The joints are not currently a part of
sympy but were sitting in a PR.

jbm950 committed:
Merge branch 'master' into featherstone

jbm950 commented:
I have added Sahil's joint code to this PR for now so I can actually start running `FeatherstoneMethod`'s code. I have removed all of the joint code/tests/docs except for a revolute and prismatic joint as those are the base of what I'm currently working on with Featherstone's method.

I have a couple of questions @moorepants 

1. Since we're moving towards interacting with a dynamic system through the `SymbolicSystem` object I don't want to mirror kane and lagrange's api where the equations of motion can be obtained from `*Method.rhs()`. I'm not sure what other methods this would leave though and I was wondering if `FeatherstonesMethod` would be better of as a function that returns a `SymbolicSystem` instance instead of a class whose only method is `.to_system()`.

2. In the joint code, why is the actual application of the joint delayed to a separate method instead of being run at the time of initialization of the joint object? Since this is the primary use of the joint object in the first place I don't see a use in making users code an additional line every time that they want to make a joint.

jbm950 commented:
In response to # 2.  It appears the method `.apply_joint()` is run at the end of `Joint.__init__()` (the base class) and is not something the user has to explicitly call to make the joint.

jbm950 commented:
Couple more questions on joints.

3. Should we default to using z axes for the joints? I'm pretty sure that's how simbody does it and every dynamics book I've looked at (thought this is not many) uses the z axis as the shared joint axis.

4. Why are parent_point_pos and child_point_pos 3 tuples instead of vectors? If they were just vectors it would eliminate the need for a `convert_tuple_to_vector` function.

jbm950 committed:
Revised and expanded the joint docstrings

I added an attributes section to the docstrings of pinjoint and
slidingjoint. I also made some minor revisions throughout the
docstrings.

jbm950 commented:
Moving all of my questions to one location for clarity and completeness and am adding a few.

1. Since we're moving towards interacting with a dynamic system through the `SymbolicSystem` object I don't want to necessarily mirror kane and lagrange's api where the equations of motion can be obtained from `*Method.rhs()`. I'm not sure what other methods this would leave though and so I was wondering if `FeatherstonesMethod` would be better off as a function that returns a `SymbolicSystem` instance instead of a class whose only method is `.to_system()`.

1. Should we default to using z-axes for the joints? I'm pretty sure that's how simbody does it and every dynamics book I've looked at (though this is not many) uses the z axis as the shared joint axis. (I've read through the entirety of the PR where @sahilshekhawat added the joints and @oliverlee mentioned using [Denavit-Hartenberg convention](https://en.wikipedia.org/wiki/Robotics_conventions#Denavit.E2.80.93Hartenberg_link_frame_convention_.28DH.29) which would agree with using z-axes for the joint axis).

1. Why are parent_point_pos and child_point_pos 3 tuples instead of vectors? If they were just vectors it would eliminate the need for a `convert_tuple_to_vector` function. What was the design reasoning behind this?

1. Another issue @oliverlee brought up was confusion in how the joint was set up. I think this is in part due to the fact that the parent body frame and child body frame at the joint location inherits its orientation from the mass center of their respective bodies.  This means that they might not be conveniently lined up with the joint axis and so a joint axis has to be supplied for both of their frames (hence the parent_axis and child_axis arguments, at least this is what I believe is what is happening). This makes the overall process somewhat complicated, and will make the joints difficult to use for defining the spatial info I need for Featherstone's method. I suggest that the frames for the two bodies at the joint location be initially aligned and a single `joint_axis` be provided. To allow this an orientation will need to be defined from the mass center frame to the joint frame for each body. I think this would make the joint formation easier to use, much clearer what is going on, and will make the spatial vector additions easier to add.

1. I think that it should be required for a user to specify coordinates for a joint rather than trying to come up with unique default values. This could be done with speeds as well or create dummy variables and set u=q'. If speeds were included I think the kinematic differential equations would need to be a possible argument as well and so I think it would be easier to default to u=q'. Would be open to thoughts on how to handle the generalized speeds. If the coordinate is a required input, would there remain any purpose in giving the joint a name that may or may not be unique? Does the joint really need to be named?

jbm950 committed:
added a function to find the spatial inertia at a specified location to the body object

jbm950 committed:
coordinate and speed are now required inputs to joints

jbm950 committed:
Changed default joint axis

I have changed the default joint axis to match denavit-hartenberg
convention and the convention used by simbody.

jbm950 committed:
Heavily revised joint.py code

I reworked the code so a lot of the redundant code is placed in
joint.__init__ instead of being repeated. I also made it so each joint
has 4 frames and 4 points (one set for each mass center and the joint
location for each body). The docstrings do not necessarily reflect the
changes yet.

jbm950 committed:
Added spatial vector information to joints

Added joint transform and motion subspace to the joints and created a
method to return these along with a spatial velocity.

moorepants commented:
Brandon, here is an example of how you can define a general pin joint:

http://nbviewer.jupyter.org/gist/moorepants/e54ecb80998a90704060b7e247ec11e8

There is a figure in the gist too:

https://gist.github.com/moorepants/e54ecb80998a90704060b7e247ec11e8

We should be able to do the same thing with the other joint types.

jbm950 commented:
Are you sending me this because what I did in rewriting the code yesterday is incorrect?

moorepants commented:
I'm sending you this because I told you I would. I haven't looked at the details of this PR yet.

jbm950 committed:
Made a bunch of changes to get joints.py to work

Made changes in multiple locations to get the joints code to work. I
think that it would work fine now though. Included in changes is a fix
to a bug in the ReferenceFrame class.

jbm950 committed:
Got everything for featherstones method running

I have the code such that FeatherstonesMethod will now run with out any
errors.

jbm950 committed:
Added some spatial vector functions

Added some functions to process matrices as spatial vectors.

jbm950 committed:
Added tests for Body.spatial_inertia()

I added tests for Body.spatial_inertia() and in doing so found errors
that I then fixed.

jbm950 committed:
Started adding tests for joint.py

This code is not yet complete and is being commited for discussion
purposes.

jbm950 committed:
Initial commit for featherstone tests and docs

This is the initial commit to add the files for featherstone tests and
documentation.

jbm950 committed:
Updating the joint code and joint test code for discussion

I am sending what I have of the joint code and joint test code for
discussion purposes.

jbm950 commented:
Here are pictures of what I'm using for the joint test code. I do not know why they loaded sideways, locally they're oriented correctly
![20160818_100230](https://cloud.githubusercontent.com/assets/12311558/17776655/6f56c736-652b-11e6-8637-991d0a28cd3f.jpg)
![20160818_100243](https://cloud.githubusercontent.com/assets/12311558/17776654/6f52b9de-652b-11e6-89cd-dca504f66ae7.jpg)
![20160818_100257](https://cloud.githubusercontent.com/assets/12311558/17776656/6f5727a8-652b-11e6-8c42-6c168a0a7c1b.jpg)
![20160818_100309](https://cloud.githubusercontent.com/assets/12311558/17776657/6f5f6418-652b-11e6-860f-d0e78b177a30.jpg)

czgdp1807 commented:
Can we add this to GSoC ideas list for next year? @moorepants

TJStienstra commented:
Just came across http://royfeatherstone.org/spatial/ and thought it might be useful to mention it here, if anyone were ever to pickup this PR.
